File: app/src/main/java/com/tealium/example/EngagementActivity.java
package com.tealium.example;

import android.os.Bundle;
import android.util.JsonReader;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatButton;

import com.tealium.example.helper.TealiumHelper;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

public class EngagementActivity extends AppCompatActivity {

    public static final String TAG = EngagementActivity.class.getSimpleName();

    EditText mFacebookUserIdEditText;
    EditText mFacebookFriendCountEditText;
    EditText mTwitterUserIdEditText;
    EditText mTwitterDescriptionEditText;

    RadioGroup mEmailSubscriptionRadioGroup;
    RadioGroup mPushSubscriptionRadioGroup;

    AppCompatButton mSaveButton;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_engagement);

        mFacebookUserIdEditText = findViewById(R.id.edit_txt_fb_username);
        mFacebookFriendCountEditText = findViewById(R.id.edit_txt_fb_friend_count);
        mTwitterUserIdEditText = findViewById(R.id.edit_txt_tw_id);
        mTwitterDescriptionEditText = findViewById(R.id.edit_txt_tw_user_description);

        mEmailSubscriptionRadioGroup = findViewById(R.id.radio_grp_email_subscription);
        mPushSubscriptionRadioGroup = findViewById(R.id.radio_grp_push_subscription);

        mSaveButton = findViewById(R.id.btn_save_engagement_details);
        mSaveButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Map<String, Object> data = getFieldData();
                TealiumHelper.trackEvent("setengagement", data);
                TealiumHelper.trackEvent("facebook", data);
                TealiumHelper.trackEvent("twitter", data);
            }
        });
    }

    private Map<String, Object> getFieldData() {
        Map<String, Object> data = new HashMap<>();

        data.put("facebook_user", mFacebookUserIdEditText.getText().toString());
        data.put("facebook_friends_count", mFacebookFriendCountEditText.getText().toString());

        data.put("twitter_user", mTwitterUserIdEditText.getText().toString());
        data.put("twitter_user_description", mTwitterDescriptionEditText.getText().toString());

        if (mEmailSubscriptionRadioGroup.getCheckedRadioButtonId() != -1) {
            data.put("email_subscription", getSubscriptionTypeFromRadioButton(mEmailSubscriptionRadioGroup.getCheckedRadioButtonId()));
        }

        if (mPushSubscriptionRadioGroup.getCheckedRadioButtonId() != -1) {
            data.put("push_subscription", getSubscriptionTypeFromRadioButton(mPushSubscriptionRadioGroup.getCheckedRadioButtonId()));
        }

        return data;
    }

    private String getSubscriptionTypeFromRadioButton(int id) {
        RadioButton selection = findViewById(id);
        if (selection.getText().toString().equals(getString(R.string.hint_opted_in))) {
            return getString(R.string.hint_opted_in_value);
        } else if (selection.getText().toString().equals(getString(R.string.hint_subscribed))) {
            return getString(R.string.hint_subscribed_value);
        } else if (selection.getText().toString().equals(getString(R.string.hint_unsubscribed))) {
            return getString(R.string.hint_unsubscribed_value);
        }
        return "";
    }


}

File: app/src/main/java/com/tealium/example/EngagementActivity.kt
package com.tealium.example

import android.os.Bundle
import android.widget.RadioButton
import android.widget.RadioGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.AppCompatButton
import com.tealium.example.helper.TealiumHelper.trackEvent

class EngagementActivity : AppCompatActivity() {
    private lateinit var emailSubscriptionRadioGroup: RadioGroup
    private lateinit var pushSubscriptionRadioGroup: RadioGroup
    private lateinit var saveButton: AppCompatButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_engagement)
        emailSubscriptionRadioGroup = findViewById(R.id.radio_grp_email_subscription)
        pushSubscriptionRadioGroup = findViewById(R.id.radio_grp_push_subscription)
        saveButton = findViewById(R.id.btn_save_engagement_details)
        saveButton.setOnClickListener {
            val data = fieldData
            trackEvent("setengagement", data)
            data["email_subscription"]?.let { emailSubscription ->
                if (emailSubscription == getString(R.string.hint_subscribed_value)) {
                    trackEvent("subscribed", mapOf("subscription_group" to "email"))
                } else if (emailSubscription == getString(R.string.hint_unsubscribed_value)) {
                    trackEvent("unsubscribed", mapOf("subscription_group" to "email"))
                }
            }

            data["push_subscription"]?.let { pushSubscription ->
                if (pushSubscription == getString(R.string.hint_subscribed_value)) {
                    trackEvent("subscribed", mapOf("subscription_group" to "push"))
                } else if (pushSubscription == getString(R.string.hint_unsubscribed_value)) {
                    trackEvent("unsubscribed", mapOf("subscription_group" to "push"))
                }
            }
        }
    }

    private val fieldData: Map<String, Any>
        get() {
            val data = mutableMapOf<String, Any>()
            if (emailSubscriptionRadioGroup.checkedRadioButtonId != -1) {
                data["email_subscription"] = getSubscriptionTypeFromRadioButton(emailSubscriptionRadioGroup.checkedRadioButtonId)
            }
            if (pushSubscriptionRadioGroup.checkedRadioButtonId != -1) {
                data["push_subscription"] = getSubscriptionTypeFromRadioButton(pushSubscriptionRadioGroup.checkedRadioButtonId)
            }
            return data
        }

    private fun getSubscriptionTypeFromRadioButton(id: Int): String {
        val selection = findViewById<RadioButton>(id)
        if (selection.text.toString() == getString(R.string.hint_opted_in)) {
            return getString(R.string.hint_opted_in_value)
        } else if (selection.text.toString() == getString(R.string.hint_subscribed)) {
            return getString(R.string.hint_subscribed_value)
        } else if (selection.text.toString() == getString(R.string.hint_unsubscribed)) {
            return getString(R.string.hint_unsubscribed_value)
        }
        return ""
    }

    companion object {
        val TAG = EngagementActivity::class.java.simpleName
    }
}